ffmpeg/ffmpeg_test.go

```diff
@@ -602,6 +604,159 @@ func TestTranscoder_MuxerOpts(t *testing.T) {
 	run(cmd)
 }
 
+type TranscodeOptionsTest struct {
+	InputCodec   VideoCodec
+	OutputCodec  VideoCodec
+	InputAccel   Acceleration
+	OutputAccel  Acceleration
+	Profile      VideoProfile
+	QualityCheck bool
+}
+
+func TestSW_Transcoding(t *testing.T) {
+	CodecsComboTest(t, supportedCodecsCombinations([]Acceleration{Software}))
+}
+
+func supportedCodecsCombinations(accels []Acceleration) []TranscodeOptionsTest {
+	prof := P240p30fps16x9
+	var opts []TranscodeOptionsTest
+	inCodecs := []VideoCodec{H264, H265, VP8, VP9}
+	outCodecs := []VideoCodec{H264, H265, VP8, VP9}
+	for _, inAccel := range accels {
+		for _, outAccel := range accels {
+			for _, inCodec := range inCodecs {
+				for _, outCodec := range outCodecs {
+					// skip unsupported combinations
+					switch outAccel {
+						case Nvidia:
+							switch outCodec {
+								case VP8, VP9:
+									continue
+								}
+					}
+					opts = append(opts, TranscodeOptionsTest{
+						InputCodec:   inCodec,
+						OutputCodec:  outCodec,
+						InputAccel:   inAccel,
+						OutputAccel:  outAccel,
+						Profile:      prof,
+						QualityCheck: true,
+					})
+				}
+			}
+		}
+	}
+	return opts
+}
+
+func CodecsComboTest(t *testing.T, options []TranscodeOptionsTest) {
+	run, dir := setupTest(t)
+	defer os.RemoveAll(dir)
+	sampleName := dir + "/test.ts"
+	var inName, outName, qName string
+	cmd := `
+    # set up initial input; truncate test.ts file
+    ffmpeg -loglevel warning -i "$1"/../transcoder/test.ts -c:a copy -c:v copy -t 1 test.ts
+  `
+	run(cmd)
+	var err error
+	for i := range options {
+		curOptions := options[i]
+		switch curOptions.InputCodec {
+			case VP8, VP9:
+				inName = dir + "/test_in.mkv"
+			case H264, H265:
+				inName = dir + "/test_in.ts"
+		}
+		switch curOptions.OutputCodec {
+			case VP8, VP9:
+				outName = dir + "/out.mkv"
+				qName = dir + "/sw.mkv"
+			case H264, H265:
+				outName = dir + "/out.ts"
+				qName = dir + "/sw.ts"
+		}
+		// if non-h264 test requested, transcode to target input codec first
+		prepare := true
+		if curOptions.InputCodec != H264 {
+			profile := P720p60fps16x9
+			profile.Encoder = curOptions.InputCodec
+			err = Transcode2(&TranscodeOptionsIn{
+				Fname: sampleName,
+				Accel: Software,
+			}, []TranscodeOptions{
+				{
+					Oname:   inName,
+					Profile: profile,
+					Accel:   Software,
+				},
+			})
+			if err != nil {
+				t.Error(err)
+				t.Fail()
```

Comment:
[t.Error](https://pkg.go.dev/testing#T.Error) is equivalent to `t.Log` followed by `t.Fail` so the `t.Fall` call should not be necessary.